nh.c

```diff
@@ -461,3 +471,50 @@ void nh_failsafe_uninstall(nh_failsafe_t *fs) {
     nh_log("(NickelHook) ... failsafe: info: deleting %s", fs->tmp);
     unlink(fs->tmp);
 }
+
+// --- File/directory deletion
+
+// The following blacklist helps prevent any whoopsies when deleting files or directories during uninstall
+
+// Let's not allow deleting any files in /bin, /sbin, /etc/init.d or u-boot stuff.
+static const char *delete_prefix_blacklist[] = {
+    "/bin",
+    "/sbin",
+    // note, '/etc/u-boot/*/u-boot.recovery' is a critical file for recovery
+    "/etc/u-boot",
+    "/etc/init.d",
+    "/usr/local/Kobo/pickel",
+};
+
+// This is for the schmuck who forgot to terminate his array of paths
+// and accidentally deleted '/bin/sh' due to an out-of-bounds read.
+void nh_delete_path(const char *path, bool is_dir) {
+    if (!path) {
+        nh_log("(NickelHook) no path supplied");
+        return;
+    }
+    char canonical_path[PATH_MAX] = {0};
+    if (realpath(path, canonical_path)) {
+        for (unsigned int i = 0; i < (sizeof(delete_prefix_blacklist) / sizeof(delete_prefix_blacklist[0])); i++) {
+            if (!strncmp(delete_prefix_blacklist[i], canonical_path, strlen(delete_prefix_blacklist[i]))) {
+                nh_log("(NickelHook) not deleting %s with blacklisted prefix %s", canonical_path, delete_prefix_blacklist[i]);
+                return;
+            }
+        }
+        nh_log("(NickelHook) ... deleting %s %s", (is_dir ? "directory" : "file") ,canonical_path);
+        int res = is_dir ? rmdir(canonical_path) : unlink(canonical_path);
+        if (res == -1) {
+            nh_log("(NickelHook) failed to delete %s, with error: %m", canonical_path);
+        }
+    } else {
+        nh_log("(NickelHook) unable to get canonical path for %s : %m", path);
+    }
+}
+
+void nh_delete_file(const char *path) {
+    nh_delete_path(path, false);
+}
+
+void nh_delete_dir(const char *path) {
+    nh_delete_path(path, true);
+}
```

Comment:
Missing LF there ;p

Reply:
Forgive me for being a bit dense this morning, but where?

Reply:
The very final line isn't terminated by a LF (hence the stop sign in the GH UI).

Unless I'm misremembering how the GH UI displays that ^^.

Reply:
Ah gotcha. I'm not gonna commit it on it's own, will be bundled into whatever the next commit is.